plugins/newsdownloader.koplugin/epubdownloadbackend.lua

```diff
@@ -0,0 +1,524 @@
+    -- But time taken by DNS lookup cannot easily be accounted for, so
+    -- a request may (when dns lookup takes time) exceed timeout and maxtime...
+    local request, sink = {}, {}
+    if maxtime then
+        request.sink = sink_table_with_maxtime(sink, maxtime)
+    else
+        request.sink = ltn12.sink.table(sink)
+    end
+    request.url = url
+    request.method = "GET"
+    local parsed = socket_url.parse(url)
+
+    local httpRequest = parsed.scheme == "http" and http.request or https.request
+    logger.dbg("request:", request)
+    local code, headers, status = socket.skip(1, httpRequest(request))
+    logger.dbg("After httpRequest")
+    local content = table.concat(sink) -- empty or content accumulated till now
+    logger.dbg("type(code):", type(code))
+    logger.dbg("code:", code)
+    logger.dbg("headers:", headers)
+    logger.dbg("status:", status)
+    logger.dbg("#content:", #content)
+
+    if code == TIMEOUT_CODE or code == MAXTIME_CODE then
+        logger.warn("request interrupted:", code)
+        return false, code
+    end
+    if headers == nil then
+        logger.warn("No HTTP headers:", code, status)
+        return false, "Network or remote server unavailable"
+    end
+    if not code or string.sub(code, 1, 1) ~= "2" then -- all 200..299 HTTP codes are OK
+        if code and code > 299 and code < 400  and headers and headers.location then -- handle 301, 302...
+           local redirected_url = headers.location
+           logger.dbg("getUrlContent: Redirecting to url: ", redirected_url)
+           return getUrlContent(redirected_url, timeout, maxtime, redirectCount + 1)
+        else
+           error("EpubDownloadBackend: Don't know how to handle HTTP response status: ", status)
+        end
+        logger.warn("HTTP status not okay:", code, status)
+        return false, "Remote server error or unavailable"
+    end
+    if headers and headers["content-length"] then
+        -- Check we really got the announced content size
+        local content_length = tonumber(headers["content-length"])
+        if #content ~= content_length then
+            return false, "Incomplete content received"
+        end
+    end
+    logger.dbg("Returning content ok")
+    return true, content
+end
+
+function EpubDownloadBackend:getResponseAsString(url)
+    logger.dbg("EpubDownloadBackend:getResponseAsString(", url, ")")
+    local success, content = getUrlContent(url)
+    if (success) then
+        return content
+    else
+        error("Failed to download content for url:", url)
+    end
+end
+
+function EpubDownloadBackend:setTrapWidget(trap_widget)
+    self.trap_widget = trap_widget
+end
+
+function EpubDownloadBackend:resetTrapWidget()
+    self.trap_widget = nil
+end
+
+function EpubDownloadBackend:loadPage(url)
+    local completed, success, content
+    if self.trap_widget then -- if previously set with Wikipedia:setTrapWidget()
```

Comment:
You should probably change this comment. ;-)

Reply:
Done :)